### DISCUSSION TITLE
Installing IfcOpenshell - include paths issue

### DISCUSSION THREAD

**Discussion by SP-AWN** [IfcOpenShell/IfcOpenShell#5468] [Q&A]
Hi,
I have successfully built and installed libraries. At least I thought so. After investigation  I have noticed that installed headers are in different folder than it is required by other library headers. eg.:  
/usr/local/include/ifcgeom_schema_agnostic/ConversionResult.h requires IfcGeomRenderStyles.h

#include "../ifcgeom/IfcGeomRenderStyles.h"

IfcGeomRenderStyles.h is not located in /usr/local/include/ifcgeom as #include directive suggest, but is also located in /usr/local/include/ifcgeom_schema_agnostic/
It is plenty of such situations. Why? What I'm doing wrong?
This is my entire install manifest:

Install the project...
-- Install configuration: "Release"
-- Installing: /usr/local/bin/IfcConvert
-- Set non-toolchain portion of runtime path of "/usr/local/bin/IfcConvert" to ""
-- Installing: /usr/local/bin/IfcGeomServer
-- Set non-toolchain portion of runtime path of "/usr/local/bin/IfcGeomServer" to ""
-- Installing: /usr/local/bin/svgfill
-- Installing: /usr/local/lib/liblibsvgfill.a
-- Installing: /usr/local/include/progress.h
-- Installing: /usr/local/include/svgfill.h
-- Installing: /usr/local/include/ifcparse/Argument.h
-- Installing: /usr/local/include/ifcparse/ArgumentType.h
-- Installing: /usr/local/include/ifcparse/IfcAlignmentHelper.h
-- Installing: /usr/local/include/ifcparse/IfcBaseClass.h
-- Installing: /usr/local/include/ifcparse/IfcCharacterDecoder.h
-- Installing: /usr/local/include/ifcparse/IfcEntityInstanceData.h
-- Installing: /usr/local/include/ifcparse/IfcException.h
-- Installing: /usr/local/include/ifcparse/IfcFile.h
-- Installing: /usr/local/include/ifcparse/IfcGlobalId.h
-- Installing: /usr/local/include/ifcparse/IfcHierarchyHelper.h
-- Installing: /usr/local/include/ifcparse/IfcLogger.h
-- Installing: /usr/local/include/ifcparse/IfcParse.h
-- Installing: /usr/local/include/ifcparse/IfcSIPrefix.h
-- Installing: /usr/local/include/ifcparse/IfcSchema.h
-- Installing: /usr/local/include/ifcparse/IfcSpfHeader.h
-- Installing: /usr/local/include/ifcparse/IfcSpfStream.h
-- Installing: /usr/local/include/ifcparse/IfcWrite.h
-- Installing: /usr/local/include/ifcparse/aggregate_of_instance.h
-- Installing: /usr/local/include/ifcparse/ifc_parse_api.h
-- Installing: /usr/local/include/ifcparse/macros.h
-- Installing: /usr/local/include/ifcparse/utils.h
-- Installing: /usr/local/include/ifcparse/variantarray.h
-- Installing: /usr/local/include/ifcparse/Ifc2x3.h
-- Installing: /usr/local/include/ifcparse/Ifc2x3-definitions.h
-- Installing: /usr/local/include/ifcparse/Ifc4.h
-- Installing: /usr/local/include/ifcparse/Ifc4-definitions.h
-- Installing: /usr/local/include/ifcparse/Ifc4x1.h
-- Installing: /usr/local/include/ifcparse/Ifc4x1-definitions.h
-- Installing: /usr/local/include/ifcparse/Ifc4x2.h
-- Installing: /usr/local/include/ifcparse/Ifc4x2-definitions.h
-- Installing: /usr/local/include/ifcparse/Ifc4x3.h
-- Installing: /usr/local/include/ifcparse/Ifc4x3-definitions.h
-- Installing: /usr/local/include/ifcparse/Ifc4x3_tc1.h
-- Installing: /usr/local/include/ifcparse/Ifc4x3_tc1-definitions.h
-- Installing: /usr/local/include/ifcparse/Ifc4x3_add1.h
-- Installing: /usr/local/include/ifcparse/Ifc4x3_add1-definitions.h
-- Installing: /usr/local/include/ifcparse/Ifc4x3_add2.h
-- Installing: /usr/local/include/ifcparse/Ifc4x3_add2-definitions.h
-- Installing: /usr/local/lib/libIfcParse.a
-- Installing: /usr/local/include/ifcgeom/mapping.h
-- Installing: /usr/local/include/ifcgeom_schema_agnostic/AbstractKernel.h
-- Installing: /usr/local/include/ifcgeom_schema_agnostic/ConversionResult.h
-- Installing: /usr/local/include/ifcgeom_schema_agnostic/ConversionSettings.h
-- Installing: /usr/local/include/ifcgeom_schema_agnostic/Converter.h
-- Installing: /usr/local/include/ifcgeom_schema_agnostic/GeometrySerializer.h
-- Installing: /usr/local/include/ifcgeom_schema_agnostic/IfcGeomElement.h
-- Installing: /usr/local/include/ifcgeom_schema_agnostic/IfcGeomFilter.h
-- Installing: /usr/local/include/ifcgeom_schema_agnostic/IfcGeomRenderStyles.h
-- Installing: /usr/local/include/ifcgeom_schema_agnostic/IfcGeomRepresentation.h
-- Installing: /usr/local/include/ifcgeom_schema_agnostic/Iterator.h
-- Installing: /usr/local/include/ifcgeom_schema_agnostic/IteratorSettings.h
-- Installing: /usr/local/include/ifcgeom_schema_agnostic/Serializer.h
-- Installing: /usr/local/include/ifcgeom_schema_agnostic/abstract_mapping.h
-- Installing: /usr/local/include/ifcgeom_schema_agnostic/empty.h
-- Installing: /usr/local/include/ifcgeom_schema_agnostic/ifc_geom_api.h
-- Installing: /usr/local/include/ifcgeom_schema_agnostic/infra_sweep_helper.h
-- Installing: /usr/local/include/ifcgeom_schema_agnostic/piecewise_function_evaluator.h
-- Installing: /usr/local/include/ifcgeom_schema_agnostic/piecewise_function_impl.h
-- Installing: /usr/local/include/ifcgeom_schema_agnostic/profile_helper.h
-- Installing: /usr/local/include/ifcgeom_schema_agnostic/taxonomy.h
-- Installing: /usr/local/lib/libgeometry_mapping_ifc2x3.a
-- Installing: /usr/local/lib/libgeometry_mapping_ifc4.a
-- Installing: /usr/local/lib/libgeometry_mapping_ifc4x1.a
-- Installing: /usr/local/lib/libgeometry_mapping_ifc4x2.a
-- Installing: /usr/local/lib/libgeometry_mapping_ifc4x3.a
-- Installing: /usr/local/lib/libgeometry_mapping_ifc4x3_tc1.a
-- Installing: /usr/local/lib/libgeometry_mapping_ifc4x3_add1.a
-- Installing: /usr/local/lib/libgeometry_mapping_ifc4x3_add2.a
-- Installing: /usr/local/lib/libgeometry_kernel_cgal.a
-- Installing: /usr/local/lib/libgeometry_kernel_cgal_simple.a
-- Installing: /usr/local/lib/libgeometry_kernel_opencascade.a
-- Installing: /usr/local/lib/libIfcGeom.a
-- Installing: /usr/local/lib/libSerializers.a
-- Installing: /usr/local/lib/libSerializers_ifc2x3.a
-- Installing: /usr/local/lib/libSerializers_ifc4.a
-- Installing: /usr/local/lib/libSerializers_ifc4x1.a
-- Installing: /usr/local/lib/libSerializers_ifc4x2.a
-- Installing: /usr/local/lib/libSerializers_ifc4x3.a
-- Installing: /usr/local/lib/libSerializers_ifc4x3_tc1.a
-- Installing: /usr/local/lib/libSerializers_ifc4x3_add1.a
-- Installing: /usr/local/lib/libSerializers_ifc4x3_add2.a
-- Installing: /usr/local/include/serializers/ColladaSerializer.h
-- Installing: /usr/local/include/serializers/GltfSerializer.h
-- Installing: /usr/local/include/serializers/HdfSerializer.h
-- Installing: /usr/local/include/serializers/IgesSerializer.h
-- Installing: /usr/local/include/serializers/OpenCascadeBasedSerializer.h
-- Installing: /usr/local/include/serializers/StepSerializer.h
-- Installing: /usr/local/include/serializers/SvgSerializer.h
-- Installing: /usr/local/include/serializers/USDSerializer.h
-- Installing: /usr/local/include/serializers/WavefrontObjSerializer.h
-- Installing: /usr/local/include/serializers/XmlSerializer.h
-- Installing: /usr/local/include/serializers/serializers_api.h
-- Installing: /usr/local/include/serializers/util.h
-- Installing: /usr/local/include/serializers/ColladaSerializer.cpp
-- Installing: /usr/local/include/serializers/GltfSerializer.cpp
-- Installing: /usr/local/include/serializers/HdfSerializer.cpp
-- Installing: /usr/local/include/serializers/OpenCascadeBasedSerializer.cpp
-- Installing: /usr/local/include/serializers/SvgSerializer.cpp
-- Installing: /usr/local/include/serializers/USDSerializer.cpp
-- Installing: /usr/local/include/serializers/WavefrontObjSerializer.cpp
-- Installing: /usr/local/include/serializers/XmlSerializer.cpp
-- Installing: /usr/local/include/serializers/util.cpp
-- Installing: /usr/local/include/serializers/schema_dependent/XmlSerializer.h
-- Installing: /usr/local/include/serializers/schema_dependent/XmlSerializer.cpp

Regards
Adam

**Comment by aothms** [comment]
Hi I think there's some version mismatch happening here. In the transition from 0.7 to 0.8 quite a bit was reshuffled. `ifcgeom_schema_agnostic` does not exist anymore, this is `ifcgeom` now. `ifcgeom` basically became `ifcgeom/kernels/opencascade` and `ifcgeom/mapping`.

We're trying our best with some form of compatibility which is why all these renames tend to a be a bit half baked. But account for big architectural changes and functionality (multiple schemas simultaneously at runtime from 0.5 to 0.6 I think). And now from 0.7 to 0.8 multiple geometry libraries (CGAL, opencascade) that can be relied on for the procedural geometry interpretation.

**Reply by SP-AWN** [reply]
Thanks.
I'm using sources from git here. Should I use some other sources? Or should I just sim-link missed paths in my system just to duck-tape it?
I'm just trying to migrate my own little library from version 0.6.0 to latest and it gives me a little bit of headache now ;) 
Regards
Adam

**Reply by SP-AWN** [reply]
> Hi I think there's some version mismatch happening here. In the transition from 0.7 to 0.8 quite a bit was reshuffled. `ifcgeom_schema_agnostic` does not exist anymore, this is `ifcgeom` now. `ifcgeom` basically became `ifcgeom/kernels/opencascade` and `ifcgeom/mapping`.
> 
I just did walk-around of this include paths problem by changing  IfcOpenshell/cmake/CMakeLists.txt.
In line 1198 have replaced SCHEMA_AGNOSTIC_H_FILES   install  destination to  DESTINATION ${INCLUDEDIR}/ifcgeom. I don't have headers linking errors. At least for now. I don't have any subfolders in ifcgeom/ 
Regards
Adam

**Comment by RickBrice** [comment]
I have run into a similar problem with v0.8 on Windows. After building the IFCOS library and building the INSTALL project (I'm using Visual Studio), I create a new C++ project. If I set the "Additional Include Directories" setting to "F:\IfcOpenShell\_installed-vs2022-x64\include" there are build errors when including certain header files. An example is

> #include <ifcgeom_schema_agnostic/abstract_mapping.h>

This file includes `../ifcgeom/taxonomy.h`. The relative path is incorrect for the INSTALLed headers.

If I set the "Additional Include Directories" to `F:\IfcOpenShell\src`, there is not an `ifcgeom_schema_agnostic` folder. I then change to 

> #include <ifcgeom/abstract_mapping.h>

and everything builds fine.

I think the root cause problem is that the folder structure in `src` is different than in `_installed-vs2022-x64\include`.

I can try to clean this up, but not certain if I have the correct understanding of the intent. Will also need a little guidance with CMake.
